cudax/include/cuda/experimental/__container/vector.h

```diff
@@ -0,0 +1,426 @@
+  using reverse_iterator       = _CUDA_VSTD::reverse_iterator<iterator>;
+  using const_reverse_iterator = _CUDA_VSTD::reverse_iterator<const_iterator>;
+  using size_type              = size_t;
+
+  //! @addtogroup construction
+  //! @{
+  //! @brief Constructs a \c vector from a memory resource
+  //! @param __mr The memory resource to allocate the vector with.
+  //! @note No storage is allocated
+  vector(_CUDA_VMR::resource_ref<_Properties...> __mr)
+      : __buf_(__mr, 0)
+      , __size_(0)
+  {}
+
+  //! @brief Constructs a \c vector of size \p __size from a memory resource and value-initializes all elements
+  //! @param __mr The memory resource to allocate the vector with.
+  //! @param __size The size of the vector.
+  vector(_CUDA_VMR::resource_ref<_Properties...> __mr, const size_t __size)
+      : __buf_(__mr, __size)
+      , __size_(__size)
+  {
+    _CUDA_VSTD::uninitialized_default_construct_n(__buf_.begin(), __size_);
+  }
+
+  //! @brief Constructs a \c vector of size \p __size from a memory resource and copy constructs its elements from
+  //! \p __value
+  //! @param __mr The memory resource to allocate the vector with.
+  //! @param __size The size of the vector.
+  //! @param __value The value all elements are copied from.
+  vector(_CUDA_VMR::resource_ref<_Properties...> __mr, const size_t __size, const _Tp& __value)
+      : __buf_(__mr, __size)
+      , __size_(__size)
+  {
+    _CUDA_VSTD::uninitialized_fill_n(__buf_.begin(), __size_, __value);
+  }
+
+  //! @brief Copy constructs a \c vector
+  //! @param __other The other vector
+  //! The new vector has capacity of \p __other.size() which is potentially less than \p __other.capacity()
+  vector(const vector& __other)
+      : __buf_(__other.resource(), __other.size())
+      , __size_(__other.size())
+  {
+    _CUDA_VSTD::uninitialized_copy(__other.begin(), __other.end(), begin());
+  }
+
+  //! @brief Move constructs a \c vector
+  //! @param __other The other vector
+  //! The new vector takes ownership of the allocation of \p __other
+  vector(vector&& __other) noexcept
+      : __buf_(_CUDA_VSTD::move(__buf_))
+      , __size_(__other.__size_)
+  {
+    __other.__size_ = 0;
+  }
+
+  //! @brief Constructs a \c vector with a given size
+  //! @param __mr The memory resource to allocate the vector with.
+  //! @param __size The size of the vector
+  //! @warning This constructor does *NOT* initialize any elements. It is the users responsibility to ensure that the
+  //! elements within `[buf.begin(), buf:begin() + size)` are properly initialized, e.g with `std::uninitialized_copy`
+  vector(_CUDA_VMR::resource_ref<_Properties...> __mr, const size_t __size, uninitialized_construction_t)
+      : __buf_(__mr, __size)
+      , __size_(__size)
+  {
+    _LIBCUDACXX_ASSERT(__size <= __buf.size(),
+                       "vector(buf, size): size is larger than the capacity of the passed buffer");
+  }
```

Comment:
there is a `cudax::uninit_t` in `__detail/utility.cuh` that i added for this purpose. also, constructor tags usually are the first parameter, no?

future musings: i can imagine there being multiple constructor tags with orthogonal meanings. they can't all be first. maybe there should be a way to combine tags, maybe with `operator|`. but that would be premature generalization. just noodling.

Reply:
Will use the uninit now that it is here